src/webgldrawer.js

```diff
@@ -0,0 +1,1235 @@
+                    let t2 = Mat3.makeTranslation(-imageCenter.x, -imageCenter.y);
+
+                    // update the view matrix to account for this image's rotation
+                    let localMatrix = t1.multiply(imageRotationMatrix).multiply(t2);
+                    overallMatrix = viewMatrix.multiply(localMatrix);
+                }
+
+                let maxTextures = this._gl.getParameter(this._gl.MAX_TEXTURE_IMAGE_UNITS);
+                let texturePositionArray = new Float32Array(maxTextures * 12); // 6 vertices (2 triangles) x 2 coordinates per vertex
+                let textureDataArray = new Array(maxTextures);
+                let matrixArray = new Array(maxTextures);
+                let opacityArray = new Array(maxTextures);
+
+                // iterate over tiles and add data for each one to the buffers
+                for(let tileIndex = 0; tileIndex < tilesToDraw.length; tileIndex++){
+                    let tile = tilesToDraw[tileIndex].tile;
+                    let index = tileIndex % maxTextures;
+                    let tileContext = tile.getCanvasContext();
+
+                    let textureInfo = tileContext ? this._TextureMap.get(tileContext.canvas) : null;
+                    if(textureInfo){
+                        this._getTileData(tile, tiledImage, textureInfo, overallMatrix, index, texturePositionArray, textureDataArray, matrixArray, opacityArray);
+                    } else {
+                        // console.log('No tile info', tile);
+                    }
+                    if( (index === maxTextures - 1) || (tileIndex === tilesToDraw.length - 1)){
+                        // We've filled up the buffers: time to draw this set of tiles
+
+                        // bind each tile's texture to the appropriate gl.TEXTURE#
+                        for(let i = 0; i <= index; i++){
+                            gl.activeTexture(gl.TEXTURE0 + i);
+                            gl.bindTexture(gl.TEXTURE_2D, textureDataArray[i]);
+                        }
+
+                        // set the buffer data for the texture coordinates to use for each tile
+                        gl.bindBuffer(gl.ARRAY_BUFFER, this._firstPass.bufferTexturePosition);
+                        gl.bufferData(gl.ARRAY_BUFFER, texturePositionArray, gl.DYNAMIC_DRAW);
+
+                        // set the transform matrix uniform for each tile
+                        matrixArray.forEach( (matrix, index) => {
+                            gl.uniformMatrix3fv(this._firstPass.uTransformMatrices[index], false, matrix);
+                        });
+                        // set the opacity uniform for each tile
+                        gl.uniform1fv(this._firstPass.uOpacities, new Float32Array(opacityArray));
+
+                        // bind vertex buffers and (re)set attributes before calling gl.drawArrays()
+                        gl.bindBuffer(gl.ARRAY_BUFFER, this._firstPass.bufferOutputPosition);
```

Comment:
> I don't think these are necessary to call every time unless the underlying data changes. OpenGL is a state machine - if something is set, it stays like that until you say otherwise (or something significant changes - like program).

You can try not setting it each time to see if it works, but setting some properties were necessary because of switching webgl programs for the two passes. I'm not totally sure how all of this works.

Reply:
> Moreover, in WebGL 2.0 you can use constant arrays so the quad buffer can be removed and replaced by a static array defined directly in the shader.

We'll have to decide if `webgl2` is the only version we want to support, I guess..

Reply:
> the texture coord is simply 0->1

The coords to sample the texture of each tile actually has to do with discarding overlap which is not needed by the webgl drawer. It doesn't have to do with tiles being off the edge of the screen, the clip space takes care of that automatically.

Reply:
>You can try not setting it each time to see if it works, but setting some properties were necessary 

I meant at least moving out most of the bindings from the for loop and doing it once at the top when the program is loaded. Also, you could probably merge both passes to a single program and avoid program switching. Not sure how much (if anything) it would save.

> The coords to sample the texture of each tile actually has to do with discarding overlap which is not needed by the webgl drawer. It doesn't have to do with tiles being off the edge of the screen, the clip space takes care of that automatically.

Yes, but the overlap is just a single number (or theoretically two at most) - a uniform could be used to compute your sampling coords on the fly and avoid sending excess stuff to the GPU. I was not talking about the clip space, just the texturing coordinates, which are in the end always between 0 and 1. The step size in the texture can be computed from the texture size (1/size.x) so the overlap can be pretty much sent to the fragment shader as-is.

Reply:
I think what you're missing is that the actual pixel overlap varies by tile - with the left-most, top-most, bottom-most and right-most tiles of each tiled image having different coordinates needed to get the right region from the texture. So you actually need to know these on a texture-by-texture basis.